sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -8999,4 +8989,211 @@ pi_result _pi_buffer::free() {
+  PI_CALL(piKernelRetain(Kernel));
+
+  ZE_CALL(zeCommandListAppendLaunchKernel,
+          (CommandBuffer->ZeCommandList, Kernel->ZeKernel,
+           &ZeThreadGroupDimensions, LaunchEvent->ZeEvent, ZeEventList.size(),
+           ZeEventList.data()));
+
+  zePrint("calling zeCommandListAppendLaunchKernel() with"
+          "  ZeEvent %#lx\n",
+          pi_cast<std::uintptr_t>(LaunchEvent->ZeEvent));
+
+  // Get sync point and register the event with it.
+  *SyncPoint = CommandBuffer->GetNextSyncPoint();
+  CommandBuffer->RegisterSyncPoint(*SyncPoint, LaunchEvent);
+  return PI_SUCCESS;
+}
+
+pi_result piextEnqueueCommandBuffer(pi_ext_command_buffer CommandBuffer,
+                                    pi_queue Queue,
+                                    pi_uint32 NumEventsInWaitList,
+                                    const pi_event *EventWaitList,
+                                    pi_event *Event) {
+
+  // Execute command list asynchronously, as the event will be used
+  // to track down its completion.
+
+  uint32_t QueueGroupOrdinal;
+  // TODO: Revisit forcing compute engine
+  auto UseCopyEngine = false;
+  auto &QGroup = Queue->getQueueGroup(UseCopyEngine);
+  auto &ZeCommandQueue =
+      // ForcedCmdQueue ? *ForcedCmdQueue :
+      QGroup.getZeQueue(&QueueGroupOrdinal);
+
+  ze_fence_handle_t ZeFence;
+  ZeStruct<ze_fence_desc_t> ZeFenceDesc;
+  pi_command_list_ptr_t CommandListPtr;
+
+  ZE_CALL(zeFenceCreate, (ZeCommandQueue, &ZeFenceDesc, &ZeFence));
+  // TODO: Refactor so requiring a map iterator is not required here, currently
+  // required for executeCommandList though.
+  std::tie(CommandListPtr, std::ignore) = CommandBuffer->CommandListMap.insert(
+      std::pair<ze_command_list_handle_t, pi_command_list_info_t>(
+          CommandBuffer->ZeCommandList,
+          {ZeFence, false, ZeCommandQueue, QueueGroupOrdinal}));
+
+  Queue->insertActiveBarriers(CommandListPtr, UseCopyEngine);
+
+  CommandListPtr->second.ZeFenceInUse = true;
+
+  // Return the command-buffer's execution event as the user visible pi_event
+  *Event = CommandBuffer->ExecutionEvent;
+  (*Event)->Queue = Queue;
+  (*Event)->RefCount.increment();
+  Queue->RefCount.increment();
+
+  PI_CALL(piEventRetain(*Event));
+
+  if (auto Res = Queue->executeCommandList(CommandListPtr, false, false))
+    return Res;
+
+  return PI_SUCCESS;
+}
+
+_pi_ext_command_buffer::_pi_ext_command_buffer(
+    pi_context Context, ze_command_list_handle_t CommandList,
+    ZeStruct<ze_command_list_desc_t> ZeDesc,
+    const pi_ext_command_buffer_desc *Desc)
+    : Context(Context), ZeCommandList(CommandList), ZeCommandListDesc(ZeDesc),
+      CommandBufferDesc(*Desc) {
+  Context->RefCount.increment();
+}
+
+_pi_ext_command_buffer::~_pi_ext_command_buffer() {
+  if (ZeCommandList) {
+    ZE_CALL_NOCHECK(zeCommandListDestroy, (ZeCommandList));
+  }
+  ExecutionEvent->RefCount.decrementAndTest();
+  Context->RefCount.decrementAndTest();
```

Comment:
Need to investigate more, but seeing an issue in a debug build here:
```bash
[ RUN      ] LevelZeroCommandBuffersTest.piextCommandBufferCreate
 #0 0x0000559daf6e42e8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (.localalias) /home/ewan/Development/dpc++/llvm/lib/Support/Unix/Signals.inc:569:22
 #1 0x0000559daf6e43a3 PrintStackTraceSignalHandler(void*) /home/ewan/Development/dpc++/llvm/lib/Support/Unix/Signals.inc:636:1
 #2 0x0000559daf6e1fae llvm::sys::RunSignalHandlers() (.localalias) /home/ewan/Development/dpc++/llvm/lib/Support/Signals.cpp:103:20
 #3 0x0000559daf6e3c20 SignalHandler(int) /home/ewan/Development/dpc++/llvm/lib/Support/Unix/Signals.inc:407:1
 #4 0x00007ff11b714420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
 #5 0x00007ff11b0cfedb std::__atomic_base<unsigned int>::operator--() /usr/include/c++/9/bits/atomic_base.h:327:34
 #6 0x00007ff11b0caff4 ReferenceCounter::decrementAndTest() /home/ewan/Development/dpc++/sycl/plugins/level_zero/pi_level_zero.hpp:312:50
 #7 0x00007ff11b0c5177 _pi_ext_command_buffer::~_pi_ext_command_buffer() /home/ewan/Development/dpc++/sycl/plugins/level_zero/pi_level_zero.cpp:9196:3
 #8 0x00007ff11b0c4326 piextCommandBufferRelease /home/ewan/Development/dpc++/sycl/plugins/level_zero/pi_level_zero.cpp:9032:10
 #9 0x0000559daf6b1e7a _pi_result sycl::_V1::detail::plugin::call_nocheck<(sycl::_V1::detail::PiApiKind)106, _pi_ext_command_buffer*>(_pi_ext_command_buffer*) const /home/ewan/Development/dpc++/sycl/source/detail/plugin.hpp:201:42
#10 0x0000559daf6a8f96 LevelZeroCommandBuffersTest_piextCommandBufferCreate_Test::TestBody() (.localalias) /home/ewan/Development/dpc++/sycl/unittests/pi/level-zero/test_ext_command_buffers.cpp:96:3
#11 0x0000559daf806d04 void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /home/ewan/Development/dpc++/llvm/utils/unittest/googletest/src/gtest.cc:2433:28
#12 0x0000559daf7ffdb8 void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /home/ewan/Development/dpc++/llvm/utils/unittest/googletest/src/gtest.cc:2488:75
#13 0x0000559daf7e0e4e testing::Test::Run() (.localalias) /home/ewan/Development/dpc++/llvm/utils/unittest/googletest/src/gtest.cc:2515:30
#14 0x0000559daf7e167b testing::TestInfo::Run() (.localalias) /home/ewan/Development/dpc++/llvm/utils/unittest/googletest/src/gtest.cc:2687:3
#15 0x0000559daf7e1d24 testing::TestSuite::Run() (.localalias) /home/ewan/Development/dpc++/llvm/utils/unittest/googletest/src/gtest.cc:2815:3
#16 0x0000559daf7ebb00 testing::internal::UnitTestImpl::RunAllTests() /home/ewan/Development/dpc++/llvm/utils/unittest/googletest/src/gtest.cc:5336:9
#17 0x0000559daf807e4e bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /home/ewan/Development/dpc++/llvm/utils/unittest/googletest/src/gtest.cc:2435:1
#18 0x0000559daf800ac5 bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /home/ewan/Development/dpc++/llvm/utils/unittest/googletest/src/gtest.cc:2488:75
#19 0x0000559daf7ea6f1 testing::UnitTest::Run() /home/ewan/Development/dpc++/llvm/utils/unittest/googletest/src/gtest.cc:4928:64
#20 0x0000559daf7d9a71 RUN_ALL_TESTS() /home/ewan/Development/dpc++/llvm/utils/unittest/googletest/include/gtest/gtest.h:2473:1
#21 0x0000559daf7d99df main /home/ewan/Development/dpc++/llvm/utils/unittest/UnitTestMain/TestMain.cpp:55:24
#22 0x00007ff11b1ae083 __libc_start_main /build/glibc-SzIz7B/glibc-2.31/csu/../csu/libc-start.c:342:3
#23 0x0000559daeefc0fe _start (/home/ewan/Development/dpc++/build_debug/tools/sycl/unittests/pi/level-zero/PiLevelZeroTests+0x220fe)
```

Reply:
Ah, I think the issue is actually the line above. Because we don't call finalize the `ExecutionEvent` is never created, and then we try to decrement it here. This patch solves it:

```diff
diff --git a/sycl/plugins/level_zero/pi_level_zero.cpp b/sycl/plugins/level_zero/pi_level_zero.cpp
index 45cbc91c0d99..987e329a72ff 100644
--- a/sycl/plugins/level_zero/pi_level_zero.cpp
+++ b/sycl/plugins/level_zero/pi_level_zero.cpp
@@ -9218,7 +9218,9 @@ _pi_ext_command_buffer::~_pi_ext_command_buffer() {
   if (ZeCommandList) {
     ZE_CALL_NOCHECK(zeCommandListDestroy, (ZeCommandList));
   }
-  ExecutionEvent->RefCount.decrementAndTest();
+  if (ExecutionEvent) {
+    ExecutionEvent->RefCount.decrementAndTest();
+  }
   Context->RefCount.decrementAndTest();
 }
```

Reply:
I've made the suggested change thanks!